tensorboard/webapp/plugins/npmi/views/annotations_list/annotation/annotation_component.ts

```diff
@@ -0,0 +1,662 @@
+      )
+      .attr(
+        'y',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.yScale(d.run)! + 15;
+        }.bind(this)
+      )
+      .attr('font-size', '13px')
+      .attr('alignment-baseline', 'middle')
+      .text(function(d: ValueData) {
+        let value =
+          d.nPMIValue === null
+            ? 'null'
+            : Math.round((d.nPMIValue + Number.EPSILON) * 1000) / 1000;
+        return value;
+      });
+
+    backgroundTexts
+      .attr(
+        'x',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.xScale(d.metric)! + 5;
+        }.bind(this)
+      )
+      .attr(
+        'y',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.yScale(d.run)! + 15;
+        }.bind(this)
+      )
+      .text(function(d: ValueData) {
+        let value =
+          d.nPMIValue === null
+            ? 'null'
+            : Math.round((d.nPMIValue + Number.EPSILON) * 1000) / 1000;
+        return value;
+      });
+
+    backgroundTexts.exit().remove();
+
+    const texts = this.textsGroup.selectAll('.npmi-text').data(this.data);
+
+    texts
+      .enter()
+      .append('text')
+      .attr('class', 'npmi-text')
+      .attr(
+        'x',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.xScale(d.metric)! + 5;
+        }.bind(this)
+      )
+      .attr(
+        'y',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.yScale(d.run)! + 15;
+        }.bind(this)
+      )
+      .attr('font-size', '13px')
+      .attr('alignment-baseline', 'middle')
+      .text(function(d: ValueData) {
+        let value =
+          d.nPMIValue === null
+            ? 'null'
+            : Math.round((d.nPMIValue + Number.EPSILON) * 1000) / 1000;
+        return value;
+      });
+
+    texts
+      .attr(
+        'x',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.xScale(d.metric)! + 5;
+        }.bind(this)
+      )
+      .attr(
+        'y',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.yScale(d.run)! + 15;
+        }.bind(this)
+      )
+      .text(function(d: ValueData) {
+        let value =
+          d.nPMIValue === null
+            ? 'null'
+            : Math.round((d.nPMIValue + Number.EPSILON) * 1000) / 1000;
+        return value;
+      });
+
+    texts.exit().remove();
+  }
+
+  private drawCountTexts() {
+    const countBackgroundTexts = this.countTextsGroup
+      .selectAll('.count-background-text')
+      .data(this.data);
+
+    countBackgroundTexts
+      .enter()
+      .append('text')
+      .attr('class', 'count-background-text')
+      .attr('stroke-width', 3)
+      .attr('stroke-linejoin', 'round')
+      .attr('stroke', 'white')
+      .attr(
+        'x',
+        function(this: AnnotationComponent, d: ValueData) {
+          return this.xScale(d.metric)! + 82;
```

Comment:
where does this number 82 come from? It looks magical :)

Reply:
Sadly, its not as magical as you might hope...
The dots are at 70, and can have r=10, which sums to 80. I dont want the text to ever touch the circle, but at the same time be at the same position for every circle, so like this it is always at least 2px away from the circle.

Is that not reasonable?

Reply:
Ah, that logic makes sense.

Can you transfer the rationale to the code with variable/constant/math?

```ts
const DOT_RADIUS = 10;
const COUNT_TEXT_OFFSET = 70;
const COUNT_TEXT_PADDING = 2;
```